File: HTB achievement/HTB.py
import aiohttp
import asyncio
import time

# Define the base URL
base_url = "https://www.hackthebox.com/achievement/machine/1694822/"

# Define the list of achievement IDs to fuzz
# Add more IDs to this list as needed
achievement_ids = ["590", "591", "592", "593", "594"]
for i in range(1, 2001):
  achievement_ids.append(str(i))

# Asynchronous function to check the validity of an achievement ID with delay
async def check_achievement(session, achievement_id, delay=1):  # Added delay argument
  # Construct the full URL
  url = base_url + achievement_id

  # Introduce a delay before sending the request
  await asyncio.sleep(delay)  # Delay for 1 second by default

  # Send an asynchronous HTTP GET request to the URL
  async with session.get(url) as response:
    # Read the response content
    content = await response.text()

    # Check for the specified messages in the response content
    if "Invalid Achievement" in content or "This achievement does not appear to be valid!" in content:
      print(f"Achievement ID {achievement_id} is invalid.")
    else:
      print(f"Achievement ID {achievement_id} is valid.")

# Main function to run the asynchronous tasks
async def main():
  # Create an asynchronous HTTP session
  async with aiohttp.ClientSession() as session:
    # Create tasks for each achievement ID
    tasks = [check_achievement(session, achievement_id) for achievement_id in achievement_ids]

    # Run the tasks concurrently with delay between each
    for task in tasks:
      await task
      await asyncio.sleep(1)  # Delay between tasks for 1 second

# Run the main function
asyncio.run(main())

File: HTB achievement/HTB.cpp
#include <iostream>
#include <future>
#include <thread>
#include <vector>
#include <string>
#include <mutex>
#include <this_thread>

// Define the base URL
const std::string base_url = "https://www.hackthebox.com/achievement/machine/1694822/";

// Function to check achievement ID validity (simulates HTTP request)
bool check_achievement(const std::string& achievement_id, int delay_ms) {
  // Simulate delay
  std::this_thread::sleep_for(std::chrono::milliseconds(delay_ms));

  // Simulate response content (replace with actual HTTP request if needed)
  std::string content = "Sample response content";

  if (content.find("Invalid Achievement") != std::string::npos ||
      content.find("This achievement does not appear to be valid!") != std::string::npos) {
    std::cout << "Achievement ID " << achievement_id << " is invalid." << std::endl;
    return false;
  } else {
    std::cout << "Achievement ID " << achievement_id << " is valid." << std::endl;
    return true;
  }
}

int main() {
  // Define achievement IDs
  std::vector<std::string> achievement_ids = {"590", "591", "592", "593", "594"};
  for (int i = 1; i < 2001; ++i) {
    achievement_ids.push_back(std::to_string(i));
  }

  // Define delay
  int delay_ms = 1000; // 1 second

  // Thread pool to manage concurrent tasks
  std::vector<std::future<bool>> results;
  std::mutex mtx; // Mutex for thread-safe output

  // Launch tasks for each achievement ID
  for (const std::string& id : achievement_ids) {
    results.push_back(std::async(std::launch::async, check_achievement, id, delay_ms));
  }

  // Wait for all tasks to finish and print results
  for (auto& result : results) {
    try {
      bool valid = result.get();
      std::lock_guard<std::mutex> lock(mtx); // Acquire lock for thread-safe output
      if (valid) {
        std::cout << "Valid ID found." << std::endl;
      }
    } catch (const std::exception& e) {
      std::cerr << "Error: " << e.what() << std::endl;
    }
  }

  return 0;
}
